File: Cpp/InformasiFilm.cpp
// =====================================================================
// FILE   : InformasiFilm.cpp
// PERAN  : BASE CLASS 1 (salah satu "orang tua" dari class Penayangan)
// ISI    : Menyimpan data yang berhubungan dengan FILM saja:
//          judul, genre, dan durasi.
// ALUR   : File ini di-include oleh Penayangan.cpp
//          -> Penayangan mewarisi class ini (multiple inheritance)
// =====================================================================

// ---- Include guard ----
// Mencegah file ini terbaca dua kali oleh compiler. Kalau tidak ada guard,
// dan file ini ter-include lewat beberapa jalur (mis. Main -> Penayangan3D
// -> Penayangan -> InformasiFilm, lalu Main -> PenayanganIMAX -> ...),
// class akan terdefinisi ganda dan terjadi error "redefinition".
#ifndef INFORMASI_FILM_CPP
#define INFORMASI_FILM_CPP

#include <string>   // dibutuhkan untuk tipe data string
using namespace std;

// ================== BASE CLASS 1 ==================
// Merepresentasikan data film yang sedang tayang di bioskop
class InformasiFilm {
    // 'protected' = atribut TIDAK bisa diakses dari luar class,
    // tetapi BISA diakses langsung oleh class turunan (Penayangan dst.).
    // Karena itu Penayangan::tampilkan() bisa memakai judulFilm, genre,
    // dan durasiMenit tanpa lewat getter.
    protected:
        string judulFilm;   // judul film, contoh: "Zootopia 2"
        string genre;       // genre film, contoh: "Animasi"
        int durasiMenit;    // lama film dalam menit, contoh: 108

    public:
        // ---- Constructor default ----
        // Dipakai saat objek dibuat tanpa argumen; semua atribut diberi
        // nilai awal kosong/0 supaya tidak berisi nilai sampah.
        InformasiFilm() {
            judulFilm = "";
            genre = "";
            durasiMenit = 0;
        }

        // ---- Constructor berparameter ----
        // Dipanggil oleh constructor Penayangan (lewat initializer list)
        // untuk mengisi data film sekaligus saat objek dibuat.
        // Akhiran 'P' pada nama parameter = "Parameter", agar tidak
        // bentrok dengan nama atribut.
        InformasiFilm(string judulFilmP, string genreP, int durasiMenitP) {
            judulFilm = judulFilmP;
            genre = genreP;
            durasiMenit = durasiMenitP;
        }

        // ---- Setter & getter ----
        // Setter : mengubah nilai atribut dari luar class.
        // Getter : membaca nilai atribut dari luar class.
        // Di Main.cpp, getter inilah yang dipakai untuk mengisi baris tabel.
        void setJudulFilm(string judulFilmP) { judulFilm = judulFilmP; }
        string getJudulFilm() { return judulFilm; }

        void setGenre(string genreP) { genre = genreP; }
        string getGenre() { return genre; }

        void setDurasiMenit(int durasiMenitP) { durasiMenit = durasiMenitP; }
        int getDurasiMenit() { return durasiMenit; }
};

#endif  // akhir dari include guard INFORMASI_FILM_CPP

File: Java/InformasiFilm.java
// =====================================================================
// FILE   : InformasiFilm.java
// PERAN  : BASE CLASS 1 (superclass dari class Penayangan)
// ISI    : Menyimpan data yang berhubungan dengan FILM saja:
//          judul, genre, dan durasi.
// ALUR   : Class ini DIWARISI oleh Penayangan (Penayangan extends InformasiFilm).
//
// CATATAN JAVA vs C++:
//   - Di Java tidak ada include guard / #include. Cukup satu class per file,
//     dan nama file HARUS sama dengan nama class public-nya
//     (InformasiFilm.java -> class InformasiFilm).
//   - Semua file .java dalam satu folder otomatis saling mengenali saat
//     dikompilasi bersama (javac *.java).
// =====================================================================

// Merepresentasikan data film yang sedang tayang di bioskop
public class InformasiFilm {

    // 'protected' = tidak bisa diakses dari luar package/class lain,
    // tetapi BISA diakses langsung oleh class turunan (Penayangan dst.).
    protected String judulFilm;   // judul film, contoh: "Zootopia 2"
    protected String genre;       // genre film, contoh: "Animasi"
    protected int durasiMenit;    // lama film dalam menit, contoh: 108

    // ---- Constructor default ----
    // Dipakai saat objek dibuat tanpa argumen; semua atribut diberi
    // nilai awal kosong/0 supaya tidak berisi null.
    public InformasiFilm() {
        this.judulFilm = "";
        this.genre = "";
        this.durasiMenit = 0;
    }

    // ---- Constructor berparameter ----
    // Dipanggil oleh constructor Penayangan lewat super(...)
    // untuk mengisi data film sekaligus saat objek dibuat.
    // 'this.judulFilm' = atribut milik objek; 'judulFilm' = parameter.
    // Dengan 'this.' nama parameter boleh sama dengan nama atribut
    // (di C++ sebelumnya dipakai akhiran 'P' untuk membedakan).
    public InformasiFilm(String judulFilm, String genre, int durasiMenit) {
        this.judulFilm = judulFilm;
        this.genre = genre;
        this.durasiMenit = durasiMenit;
    }

    // ---- Setter & getter ----
    // Setter : mengubah nilai atribut dari luar class.
    // Getter : membaca nilai atribut dari luar class.
    // Di Main.java, getter inilah yang dipakai untuk mengisi baris tabel.
    public void setJudulFilm(String judulFilm) { this.judulFilm = judulFilm; }
    public String getJudulFilm() { return judulFilm; }

    public void setGenre(String genre) { this.genre = genre; }
    public String getGenre() { return genre; }

    public void setDurasiMenit(int durasiMenit) { this.durasiMenit = durasiMenit; }
    public int getDurasiMenit() { return durasiMenit; }
}
